File: targets/js-ts-express/services/fileService.ts
// File retrieval logic.
import * as fs from 'fs';
import * as path from 'path';

const UPLOAD_DIR = '/var/www/uploads';
const ALLOWED = new Set(['readme.txt', 'help.txt']);

// Passes user-controlled name into a filesystem path (sink).
export function readUserFile(name: string): string {
  const p = path.join(UPLOAD_DIR, name);
  return fs.readFileSync(p, 'utf-8');
}

export function readWhitelisted(name: string): string {
  if (!ALLOWED.has(name)) {
    throw new Error('file not allowed');
  }
  return fs.readFileSync(path.join(UPLOAD_DIR, name), 'utf-8');
}

File: targets/jsp-legacy/src/legacy/service/FileService.java
package legacy.service;

import java.util.Set;

import legacy.config.AppConfig;
import legacy.util.FileUtil;

/** File retrieval logic. */
public class FileService {

    private static final Set<String> ALLOWED = Set.of("readme.txt", "help.txt");

    // VULN: jsp-path-traversal-01 (path-traversal, cwe-22) - passes user-controlled name down to the filesystem helper (sink hop)
    public String readUserFile(String name) throws Exception {
        return FileUtil.read(AppConfig.UPLOAD_DIR, name);
    }

    // SAFE: jsp-safe-04 (mimics path-traversal) - allow-list validation before reading
    public String readWhitelisted(String name) throws Exception {
        if (!ALLOWED.contains(name)) {
            throw new IllegalArgumentException("file not allowed");
        }
        return FileUtil.read(AppConfig.UPLOAD_DIR, name);
    }
}
